### ISSUE TITLE
[REVIEW]: DataAssimilationBenchmarks.jl: a data assimilation research framework

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#4129]
**Submitting author:** @cgrudz (<a href="http://orcid.org/0000-0002-3084-3178">Colin Grudzien</a>)
**Repository:** https://github.com/cgrudz/DataAssimilationBenchmarks.jl
**Branch with paper.md** (empty if default branch): 
**Version:** v0.3.4
**Editor:** @diehlpk
**Reviewers:** @peanutfun, @tmigot
**Archive:** 10.5281/zenodo.7311847

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736/status.svg)](https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736"><img src="https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736/status.svg)](https://joss.theoj.org/papers/478dcc0b1608d2a4d8c930edebb58736)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/cgrudz/DataAssimilationBenchmarks.jl" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@peanutfun & @tmigot, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @taless474 know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @peanutfun
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/cgrudz/DataAssimilationBenchmarks.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cgrudz/DataAssimilationBenchmarks.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@cgrudz) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @tmigot
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/cgrudz/DataAssimilationBenchmarks.jl">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cgrudz/DataAssimilationBenchmarks.jl">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@cgrudz) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @peanutfun, @tmigot it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1301

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.5281/zenodo.2029296 is OK

MISSING DOIs

- 10.1175/2009bams2618.1 may be a valid DOI for title: The data assimilation research testbed: A community facility
- 10.5194/gmd-2021-306 may be a valid DOI for title: A fast, single-iteration ensemble Kalman smoother for sequential data assimilation
- 10.1080/16000870.2018.1445364 may be a valid DOI for title: State-of-the-art stochastic data assimilation methods for high-dimensional non-Gaussian problems
- 10.5194/gmd-13-1903-2020 may be a valid DOI for title: On the numerical integration of the Lorenz-96 model, with scalar additive noise, for benchmark twin experiments
- 10.1002/wcc.535 may be a valid DOI for title: Data assimilation in the geosciences: An overview of methods, issues, and perspectives
- 10.5194/gmd-13-1903-2020 may be a valid DOI for title: On the numerical integration of the Lorenz-96 model, with scalar additive noise, for benchmark twin experiments

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.24 s (261.9 files/s, 74008.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Julia                           24           1353           3628           6036
Python                          27            971            120           3984
Markdown                         3            119              0            583
TOML                             2            108              1            451
TeX                              2             11              0            107
YAML                             3              3              4             39
Bourne Shell                     1              1              0              2
-------------------------------------------------------------------------------
SUM:                            62           2566           3753          11202
-------------------------------------------------------------------------------


Statistical information for the repository '01771c2f9ef55a9d0f8f5c70' was
gathered on 2022/02/03.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Colin Grudzien                  18          1572            312           26.23
Colin J Grudzien                 4            32            124            2.17
cgrudz                           7          2460            448           40.48
plinx                           17          2065            170           31.12

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Colin J Grudzien              7           21.9          8.4                0.00
plinx                      5068          245.4          4.4                2.35
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.04129/joss.04129/10.21105.joss.04129.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.04129/joss.04129/10.21105.joss.04129.pdf) :page_facing_up: :point_left:

**Comment by peanutfun** [comment]
Hi @cgrudz 👋 I'll review your software in the next couple of days, and open up issues in the repository as I go along. Some of them might only be suggestions and not crucial to the review. Once I'm finished, I'll report back here and summarize the issues with respect to the review criteria and the checklist above.

Happy coding! 🚀

**Comment by cgrudz** [comment]
Hi @peanutfun, thanks so much, I'll be looking forward to hearing your recommendations for improvement.

Cheers,
Colin

**Comment by cgrudz** [comment]
@peanutfun, thanks so much for these useful and detailed suggestions for improvement.  I'll start working through them slowly as I'm also in a somewhat reduced service mode with this project, as I just started a new position.  Cheers!

**Comment by peanutfun** [comment]
@cgrudz, no worries, take your time. Then I can also take a more relaxed pace with the review 😇

**Comment by whedon** [comment]
:wave: @tmigot, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @peanutfun, please update us on how your review is going (this is an automated reminder).

**Comment by peanutfun** [comment]
@cgrudz, I finished my review. Let me summarize:

DataAssimilationBenchmarks.jl is a well-working framework for researching data assimilation techniques. Installation is easy and I experienced no issues when executing functions of this package. The automated tests are working well and help understanding the inner workings of the module, although code coverage is still a bit low. The `README.md` gives a good overview of the framework and extensively documents the public API. Finally, the paper is well written and does an especially good job in motivating the development of the framework.

My main issue with the current state of the package is that I, as a person unfamiliar with the source code, see no clear way of extending the framework with new models and DA methods. In that regard, I feel that DataAssimilationBenchmarks.jl fails its own premise because it is intended to compare DA methods against each other and include more filters and smoothers in the future. As an open-source project, it should encourage contributions by others, and therefore needs instructions on how to add such new methods to the framework.

The other points I want to address mostly concern documentation. The `README.md` should include an *Example Usage* section, that illustrates the capabilities of the framework and the intended workflow, and supplies a useful set of default function arguments. As the module functions do not return data structures but write all data into files, the structure and content of these files should be documented as well. Additionally, the repository lacks community guidelines, which is more of a technicality. And finally, I think the paper can be shortened, and some of its references can be improved.

I suggested a lot more in the issues I created during my review, which is mainly due to the fact that I see a lot of potential in DataAssimilationBenchmarks.jl. But most of them are well beyond the scope of this review. Two things I still would like to point out: First, you already documented the API extensively in the `README.md`. Moving this documentation over to docstrings in the code would yield a documentation that is formatted in a standardized and expected way, and should not be much work. And second, I think that reworking the analysis scripts to work on any system would be a huge selling point. Ideally, the framework would then cover everything from running a model, over benchmarking DA methods in parallel, to evaluating and analyzing their performance.

## Issues for meeting review criteria

- [ ] cgrudz/DataAssimilationBenchmarks.jl#7
- [ ] cgrudz/DataAssimilationBenchmarks.jl#10
- [ ] cgrudz/DataAssimilationBenchmarks.jl#14
- [ ] cgrudz/DataAssimilationBenchmarks.jl#15
- [ ] Incorporate suggestions to JOSS paper (or conclusively reject them) cgrudz/DataAssimilationBenchmarks.jl#16

## Possible improvements I strongly recommend

*(These are personal suggestions to improve the package, and they are not required to meet the review criteria for publication in JOSS)*

- [ ] Use docstrings to document API cgrudz/DataAssimilationBenchmarks.jl#8
- [ ] Rework analysis scripts to be used with every experiment/benchmark cgrudz/DataAssimilationBenchmarks.jl#11

**Comment by cgrudz** [comment]
@peanutfun thank you for your excellent suggestions and detailed review, I really appreciate the effort you took to go through the code as you did.  I just wanted to write a quick message to let you know that I am intending to start on revisions and reading through the posted issues in about a month or two -- I am waiting in part to see if I can fund an undergraduate research assistant to help resolve the issues, as I think these will make good exercises for an RA again.  However, if this funding does not go through, I do intend to resolve these issues myself.  I'll keep the review process updated with my status shortly.

Cheers!
Colin

**Comment by tmigot** [comment]
Hi @cgrudz ! Sorry, I am a bit late in the review. But, I saw @peanutfun already did a great job.
I am checking the  `Functionality` section.
- Installation: Does installation proceed as outlined in the documentation?

Everything works following your suggestions. However, when adding your package the "usual" way in Julia, i.e.
```
add DataAssimilationBenchmark
```
I get the following error when testing
```
     Testing Running tests...
Test Summary:               | Pass  Total
Calculate Order Convergence |    2      2
Test Summary: | Pass  Total
Lorenz-96     |    2      2
Runtime 0.0925 minutes
┌ Warning: Opening file with JLD2.MmapIO failed, falling back to IOStream
└ @ JLD2 ~/.julia/packages/JLD2/k9Gt0/src/JLD2.jl:233
Error encountered while save FileIO.File{FileIO.DataFormat{:JLD2}, String}("/home/tmigot/.julia/packages/DataAssimilationBenchmarks/ZFBxn/src/experiments/../data/time_series/IEEE39bus_time_series_seed_0000_diff_0.000_tanl_0.01_nanl_05000_spin_1500_h_0.010.jld2").

Fatal error:
Time Series Generation: Test Failed at /home/tmigot/.julia/packages/DataAssimilationBenchmarks/ZFBxn/test/runtests.jl:41
  Expression: TestTimeSeriesGeneration.testGenIEEE39bus()
Stacktrace:
 [1] macro expansion
   @ ~/packages/julias/julia-1.7.1/share/julia/stdlib/v1.7/Test/src/Test.jl:445 [inlined]
 [2] macro expansion
   @ ~/.julia/packages/DataAssimilationBenchmarks/ZFBxn/test/runtests.jl:41 [inlined]
 [3] macro expansion
   @ ~/packages/julias/julia-1.7.1/share/julia/stdlib/v1.7/Test/src/Test.jl:1283 [inlined]
 [4] top-level scope
   @ ~/.julia/packages/DataAssimilationBenchmarks/ZFBxn/test/runtests.jl:39
Test Summary:          | Pass  Fail  Total
Time Series Generation |    3     1      4
ERROR: LoadError: Some tests did not pass: 3 passed, 1 failed, 0 errored, 0 broken.
in expression starting at /home/tmigot/.julia/packages/DataAssimilationBenchmarks/ZFBxn/test/runtests.jl:2
ERROR: Package DataAssimilationBenchmarks errored during testing
```
Maybe the error comes from Julia 1.7 or is it an issue with the `add` ?

In the latter case, this might be inconvenient, for instance if someone uses this package as a dependency. What do you think? Once we clarify this, I can open an issue on this topic.

- Functionality: Have the functional claims of the software been confirmed?

As mentioned in https://github.com/cgrudz/DataAssimilationBenchmarks.jl/issues/7  the documentation and the paper lacks an illustration of what the package is capable of doing.

**Comment by tmigot** [comment]
@cgrudz Have you thought about making a real documentation hosted with Github Pages and using `Documenter.jl`?

See docs for Documenting in Julia: https://juliadocs.github.io/Documenter.jl/stable/

I have a package that has been recently published in JOSS as an example: https://github.com/JuliaSmoothOptimizers/DCISolver.jl (you can access the doc via the badges dev and stable in the readme).

This would allow clearer and more complete documentation. That could be very relevant, especially, in the perspective of expanding your package.

**Comment by cgrudz** [comment]
@tmigot, no worries, thank you very much for your helpful suggestions and for catching the bug with the package add.  As I mentioned above I will start to work on these revisions in earnest in one to two months, once I determine the availability of a research assistant to help with these tasks.  I'll address these items as soon as I can. Cheers!

**Comment by taless474** [comment]
@cgrudz how is the development process? Did you have the chance to complete what the reviewers asked for?

**Comment by cgrudz** [comment]
@taless474 , thanks for checking in.  Things are currently in-progress where I am about ready to start clearing some of the issues ([new doc pages](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/)!), and I am expecting to have a summer research assistant who will be working with the software and will help with the revision and new features for the next release.  I am anticipating resolving the major items and submitting a revision after the research internship has completed around the end of summer, but I'll post updates as I feel like issues / tickets have been cleared.  Cheers!

**Comment by taless474** [comment]
@cgrudz Thank you for the update and good luck

**Comment by peanutfun** [comment]
@taless474 @cgrudz I am happy to see things are moving along here. Please note that due to the death of a close family member I will need to take some time off. I will continue the review on 19 September at the earliest. Thanks in advance for your understanding!

**Comment by cgrudz** [comment]
@peanutfun, I'm so sorry to hear this, you have my condolences and full understanding.  We will try to have this in re-submission in the coming few weeks, there is no rush to review this.

**Comment by arfon** [comment]
@editorialbot assign @diehlpk as editor

:wave: folks – @diehlpk has kindly volunteered to step in as the handling editor here as @taless474 is not currently available to edit. Thanks Patrick!

**Comment by diehlpk** [comment]
Hi @tmigot how is your review going?

**Comment by diehlpk** [comment]
Hi @peanutfun, I hope things are going better for you. Could you please update us on your review progress?

**Comment by diehlpk** [comment]
> Hi @diehlpk, thanks for checking in. Yes, I am better! I understood the comment by @cgrudz that he will notify us once he thinks all issues related to this review are resolved. Therefore I put the review on hold. Did I get this wrong? I can resume working on this review at any time.

@cgrudz can you please comment on that?

**Comment by cgrudz** [comment]
@diehlpk, thanks for reaching out, the timing is very good actually because we were just waiting to finish off some last test cases and push a new tagged version to the Julia General Registries.  I think everything is now complete with our revision, and the new tag is in-progress with the Julia tagbot action pending:

https://github.com/JuliaRegistries/General/pull/70433

Please feel free to start the next round of review, we are now satisfied with how we've addressed the original issues and how we have improved the quality of the software and its documentation.

Cheers,
Colin

**Comment by diehlpk** [comment]
@peanutfun can you please have a look?

**Comment by diehlpk** [comment]
@peanutfun can you please have a look?

**Comment by diehlpk** [comment]
@tmigot how is your review going?

**Comment by tmigot** [comment]
Hi @diehlpk , 
I was in the same situation described by @peanutfun waiting for some news for months, so I am unsure why you are pressing the reviewers. 
Anyway, I will get back to look at the new changes made to the package.

**Comment by tmigot** [comment]
@cgrudz I have a quick question. Is there a reason why you promote the `master` branch in the documentation and installation guidelines and not the stable release?

**Comment by tmigot** [comment]
@whedon generate pdf

**Comment by tmigot** [comment]
@editorialbot generate pdf

**Comment by diehlpk** [comment]
@cgrudz Please check the error message from here https://github.com/openjournals/joss-papers/actions/runs/3305148863/jobs/5455018587 and fix your paper.

**Comment by diehlpk** [comment]
> Hi @diehlpk , I was in the same situation described by @peanutfun waiting for some news for months, so I am unsure why you are pressing the reviewers. Anyway, I will get back to look at the new changes made to the package.

Do not wanted to pressure you, sorry for that. Just checking in since I am the new editor.

**Comment by cgrudz** [comment]
> @cgrudz Please check the error message from here https://github.com/openjournals/joss-papers/actions/runs/3305148863/jobs/5455018587 and fix your paper.

@diehlpk, I think I caught the issue, can you reinitialize a build to test?

**Comment by cgrudz** [comment]
> @cgrudz I have a quick question. Is there a reason why you promote the `master` branch in the documentation and installation guidelines and not the stable release?

@tmigot, currently Documenter.jl is configured in the project to automatically update the documentation when there is a commit to master. We are anticipating that the master branch and documentation will be kept consistent with each other, and as this is in continuous development, the master branch will be running ahead of the last tagged stable version.

**Comment by diehlpk** [comment]
@editorialbot generate pdf

**Comment by tmigot** [comment]
@cgrudz you can also visualize the pdf here https://github.com/cgrudz/DataAssimilationBenchmarks.jl/actions/workflows/draft-pdf.yml in the Artifacts of each build

**Comment by cgrudz** [comment]
Thanks for the information,

@editorialbot generate pdf

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by cgrudz** [comment]
@diehlpk, I think the article build  is now resolved, thanks for the help

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by cgrudz** [comment]
@peanutfun, thanks for the fixes to the DOIs, these are now included in the new PDF.

**Comment by peanutfun** [comment]
@diehlpk @cgrudz I concluded my review and recommend DataAssimilationBenchmarks.jl for publication. I am happy to see so many improvements since the start of this review. The new documentation page in particular is a great addition. Thank you for the fun and constructive review process!

**Comment by cgrudz** [comment]
@peanutfun thanks so much again for the thorough and extremely useful review, the suggestions (and even providing a PR) were very helpful.

**Comment by diehlpk** [comment]
@tmigot Can you please have a look at the the changes too?

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by tmigot** [comment]
@cgrudz @diehlpk I finished my review. The package looks very nice now, congrats. I still have a couple of comments.

- [ ] As asked above, https://github.com/openjournals/joss-reviews/issues/4129#issuecomment-1287949731 , I don't think it is a good idea to only focus on the master simply because there is a reason we use versioning. The user needs to rely on versions that are stable and working especially since the package has a lot of deps. Staying up-to-date with the master is great but only for developers that are actively using the package.

  The following two would solve the issue:
  - [ ] Add link to `dev` and `stable` docs to the readme https://github.com/cgrudz/DataAssimilationBenchmarks.jl/issues/29
  - [ ] Show how to install the stable version of the package
- [ ] Regarding the automated tests, it passes on my computer. However, I am slightly concerned about the code coverage as only 40% of the code is tested. This should be increased to a more satisfying level.
- [ ] Finally, I think the package still lacks a clear tutorial with a complete test scenario.

**Comment by cgrudz** [comment]
@tmigot, thanks for your suggestions.  I have resolved the issues that were raised in the main repository and I feel like the other items are now resolved as well.  @diehlpk, I would like to ask for your decision of the work in its current state.

@tmigot regarding your requests:

"Show how to install the stable version of the package"

This is performed by installing the latest stable official release in the Julia General Registries.  Please find the instructions [here](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Install-a-dev-package).

"Regarding the automated tests, it passes on my computer. However, I am slightly concerned about the code coverage as only 40% of the code is tested. This should be increased to a more satisfying level."

Please note that I have included additional test now to verify that the parallel experiment constructors work.  This has raised the coverage already, but also please not that the [Analysis Directory](https://app.codecov.io/gh/cgrudz/DataAssimilationBenchmarks.jl/tree/master/src/analysis) contains 900 tracked lines, none of which are officially supported at this time as discussed in the documentation.  If you consider the coverage of lines (1670) relative to the total tracked lines (3181) minus the non-supported Analysis codes (900), the current coverage is closer to 70% of the supported code.  I could always remove these non-supported lines, but the plan is to eventually integrate them in native Julia, though for now I feel like it doesn't hurt anyone to have a template for performing their own analysis.

"Finally, I think the package still lacks a clear tutorial with a complete test scenario."

I feel like this is adequately covered in the documentation between the discussion of the basic workflow

https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/submodules/experiments/Workflow/

and the discussion of the parallel experiment driver for running a sensitivity test

https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/submodules/experiments/ParallelExperimentDriver/#ParallelExperimentDriver

The kinds of experiments that can be run are furthermore clearly discussed in the open-access manuscript this software supported:

https://gmd.copernicus.org/articles/15/7641/2022/gmd-15-7641-2022.html

Between the documentation itself, and its works referenced, I think the usage of the software is adequate to run an analysis.

**Comment by diehlpk** [comment]
> "Regarding the automated tests, it passes on my computer. However, I am slightly concerned about the code coverage as only 40% of the code is tested. This should be increased to a more satisfying level."
> 
> Please note that I have included additional test now to verify that the parallel experiment constructors work. This has raised the coverage already, but also please not that the [Analysis Directory](https://app.codecov.io/gh/cgrudz/DataAssimilationBenchmarks.jl/tree/master/src/analysis) contains 900 tracked lines, none of which are officially supported at this time as discussed in the documentation. If you consider the coverage of lines (1670) relative to the total tracked lines (3181) minus the non-supported Analysis codes (900), the current coverage is closer to 70% of the supported code. I could always remove these non-supported lines, but the plan is to eventually integrate them, and for now I feel like it doesn't hurt anyone to have a template for performing their own analysis.

@cgrudz I do not know Julia well enough, but could you just exclude the one file from the analysis?

**Comment by diehlpk** [comment]
> This is performed by installing the latest stable official release in the Julia General Registries. Please find the instructions [here](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Install-a-dev-package).

@cgrudz I think the information is good, but I would make it more prominent. Maybe with a own header? I think many people, who want to use the package and not add features might prefer the stable version. I think making it easier to find how to do that would benefit your users.

**Comment by diehlpk** [comment]
> "Finally, I think the package still lacks a clear tutorial with a complete test scenario."

I think it is ok what you have in the documentation. However, a code example would be nice. So your users can adapt these examples do their needs.

**Comment by cgrudz** [comment]
> > "Regarding the automated tests, it passes on my computer. However, I am slightly concerned about the code coverage as only 40% of the code is tested. This should be increased to a more satisfying level."
> > Please note that I have included additional test now to verify that the parallel experiment constructors work. This has raised the coverage already, but also please not that the [Analysis Directory](https://app.codecov.io/gh/cgrudz/DataAssimilationBenchmarks.jl/tree/master/src/analysis) contains 900 tracked lines, none of which are officially supported at this time as discussed in the documentation. If you consider the coverage of lines (1670) relative to the total tracked lines (3181) minus the non-supported Analysis codes (900), the current coverage is closer to 70% of the supported code. I could always remove these non-supported lines, but the plan is to eventually integrate them, and for now I feel like it doesn't hurt anyone to have a template for performing their own analysis.
> 
> @cgrudz I do not know Julia well enough, but could you just exclude the one file from the analysis?

@diehlpk, I have removed most of the deprecated example scripts from the file, now bringing the coverage to above 70%, I hope this is satisfactory.

**Comment by diehlpk** [comment]
> > > "Regarding the automated tests, it passes on my computer. However, I am slightly concerned about the code coverage as only 40% of the code is tested. This should be increased to a more satisfying level."
> > > Please note that I have included additional test now to verify that the parallel experiment constructors work. This has raised the coverage already, but also please not that the [Analysis Directory](https://app.codecov.io/gh/cgrudz/DataAssimilationBenchmarks.jl/tree/master/src/analysis) contains 900 tracked lines, none of which are officially supported at this time as discussed in the documentation. If you consider the coverage of lines (1670) relative to the total tracked lines (3181) minus the non-supported Analysis codes (900), the current coverage is closer to 70% of the supported code. I could always remove these non-supported lines, but the plan is to eventually integrate them, and for now I feel like it doesn't hurt anyone to have a template for performing their own analysis.
> > 
> > 
> > @cgrudz I do not know Julia well enough, but could you just exclude the one file from the analysis?
> 
> @diehlpk, I have removed most of the deprecated example scripts from the file, now bringing the coverage to above 70%, I hope this is satisfactory.

Thanks. I think that looks much better.

**Comment by diehlpk** [comment]
@tmigot Do you agree with 70%?

**Comment by cgrudz** [comment]
> > This is performed by installing the latest stable official release in the Julia General Registries. Please find the instructions [here](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Install-a-dev-package).
> 
> @cgrudz I think the information is good, but I would make it more prominent. Maybe with a own header? I think many people, who want to use the package and not add features might prefer the stable version. I think making it easier to find how to do that would benefit your users.

@diehlpk, thanks for the suggestions, I have implemented the [further headings](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Installation), and I think I should clarify the scope of the package.  The use case for the software is developing the dev package for research purposes, where the standalone methods have very little functionality by themselves.  New experiments are to be constructed by directly editing the source, using the existing `ParallelExperimentDrivers.jl` and `SingleExperimentDrivers.jl` as templates.  This is also why there isn't much integration currently with the analysis scripts, as users are expected to analyze their data according to experimental parameters that they design.

**Comment by cgrudz** [comment]
> > "Finally, I think the package still lacks a clear tutorial with a complete test scenario."
> 
> I think it is ok what you have in the documentation. However, a code example would be nice. So your users can adapt these examples do their needs.

@diehlpk, I have [copied source code into the documentation](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/submodules/experiments/ParallelExperimentDriver/) directly from the `ParallelExperimentDriver.jl` as discussed above to illustrate how an experiment is run.  The expectation again is that a user needs to define the parameters of interest for themself and to create a similar constructor function to generate the NamedTuples that define an experimental configuration.  If an experiment lies outside of the scope of the existing defined experiments, a user furthermore needs to define their own experiment using the dev package, but can utilize all of the backend of the framework.

**Comment by tmigot** [comment]
> @tmigot Do you agree with 70%?

It is better to have the available code tested. This is good now, thanks!

**Comment by diehlpk** [comment]
> > > "Finally, I think the package still lacks a clear tutorial with a complete test scenario."
> > 
> > I think it is ok what you have in the documentation. However, a code example would be nice. So your users can adapt these examples do their needs.
> 
> @diehlpk, I have [copied source code into the documentation](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/submodules/experiments/ParallelExperimentDriver/) directly from the `ParallelExperimentDriver.jl` as discussed above to illustrate how an experiment is run.  The expectation again is that a user needs to define the parameters of interest for themself and to create a similar constructor function to generate the NamedTuples that define an experimental configuration.  If an experiment lies outside of the scope of the existing defined experiments, a user furthermore needs to define their own experiment using the dev package, but can utilize all of the backend of the framework.

I think that improved the concerns. @tmigot  please have a look?

**Comment by diehlpk** [comment]
> > > This is performed by installing the latest stable official release in the Julia General Registries. Please find the instructions [here](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Install-a-dev-package).
> > 
> > @cgrudz I think the information is good, but I would make it more prominent. Maybe with a own header? I think many people, who want to use the package and not add features might prefer the stable version. I think making it easier to find how to do that would benefit your users.
> 
> @diehlpk, thanks for the suggestions, I have implemented the [further headings](https://cgrudz.github.io/DataAssimilationBenchmarks.jl/dev/home/Getting%20Started/#Installation), and I think I should clarify the scope of the package.  The use case for the software is developing the dev package for research purposes, where the standalone methods have very little functionality by themselves.  New experiments are to be constructed by directly editing the source, using the existing `ParallelExperimentDrivers.jl` and `SingleExperimentDrivers.jl` as templates.  This is also why there isn't much integration currently with the analysis scripts, as users are expected to analyze their data according to experimental parameters that they design.

Please make sure to emphasize this in the paper and documentation.

**Comment by tmigot** [comment]
It looks great now, thanks.
Final comments: 
- I open a PR to fix the link to the stable doc.
- I assume there is another typo here https://github.com/cgrudz/DataAssimilationBenchmarks.jl/blob/eb357ecaade4d10f6ed23e62f777788af1970757/docs/src/home/Getting%20Started.md?plain=1#L34 where it should be `pkg> add DataAssimilationBenchmarks`?

 I can recommend DataAssimilationBenchmarks.jl for publication. Congratulation for the package, and the many improvements.

**Comment by cgrudz** [comment]
@tmigot, thanks so much for the PR's for the issue fixes, we appreciate the attention to detail with this.  I've merged the PR but I've had a busy morning and I'll push the fixes to the stable General Registries version of the repository shortly.  Likewise, thanks for your approval of the software, this was an excellent learning experience for myself and my collaborators that vastly improved our software.  We'll be looking forward to continuing our development and extending this as the framework for further research manuscripts.

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by cgrudz** [comment]
@diehlpk, please note we also added you to the acknowledgement section, thank you for your suggestions and for handling the manuscript in this phase of the review.

**Comment by diehlpk** [comment]
@cgrudz Thank you. I am traveling this weekend and will finish the editorial tasks by Tuesday. I think we can publish the paper at the end of next week.

**Comment by diehlpk** [comment]
@whedon commands

**Comment by diehlpk** [comment]
@editorialbot commands

**Comment by diehlpk** [comment]
@peanutfun and @tmigot thank you so much for your effort, we could not run JOSS without you.

**Comment by diehlpk** [comment]
@editorialbot generate pdf

**Comment by diehlpk** [comment]
@cgrudz 

- [ ] Please add the country to all affiliations.

After that 

- [ ] Generate a new release including all the changes, you made during the review and post the version here
- [ ] Upload this version to Zenodo and post the DOI here. Note that the title and the authors on Zenodo need to match with the paper.

**Comment by cgrudz** [comment]
@diehlpk that sounds good, I'm also on travel currently through Wednesday, but I should be able to handle these items on Thursday this week.  No worries with the holiday if we can't finish by Friday, I'll try to get back to this as soon as I can.

Cheers,
Colin

**Comment by cgrudz** [comment]
@editorialbot generate pdf

**Comment by cgrudz** [comment]
@diehlpk, the paper version above includes countries in the affiliations.  Please find below the new release and the associated Zenodo release

https://github.com/cgrudz/DataAssimilationBenchmarks.jl/releases/tag/v0.3.4

10.5281/zenodo.7311847

Let me know if you need any further actions and thanks for your assistance!

**Comment by diehlpk** [comment]
@editorialbot set v0.3.4 as version

**Comment by diehlpk** [comment]
@editorialbot set 10.5281/zenodo.7311847 as archive

**Comment by diehlpk** [comment]
@cgrudz The title of the Zenodo archive reads as `cgrudz/DataAssimilationBenchmarks.jl: JOSS Paper Version`, but it should read as `DataAssimilationBenchmarks.jl: a data assimilation research framework` to match the paper. Please edit the title. There is no need for a new verison.

**Comment by diehlpk** [comment]
@editorialbot check references

**Comment by diehlpk** [comment]
@editorialbot generate pdf

**Comment by diehlpk** [comment]
@cgrudz After that change, I can recommend the paper for acceptance.

**Comment by cgrudz** [comment]
Great! I think I just got this fixed, let me know if there is anything else to do.

**Comment by diehlpk** [comment]
@editorialbot recommend-accept

**Comment by diehlpk** [comment]
@cgrudz Thanks, I can now recommend the paper for acceptance.

**Comment by kthyng** [comment]
Everything looks good!

**Comment by kthyng** [comment]
@editorialbot accept

**Comment by kthyng** [comment]
Congratulations on your new publication @cgrudz!! Many thanks to editor @diehlpk and reviewers @peanutfun and @tmigot for your time, hard work, and expertise!!
